### PR TITLE
Don't issue no-op txn on settings write

### PR DESCRIPTION
**Pull request by wupeka**
## Description of change
Don't issue txn when writing settings would be a no-op

## QA steps
Bootstrap juju, deploy large bundle, verify that there're no empty txns - connect to mongo and issue db.txns.find({o:[]}).count() - it should be 0.

## Bug reference
https://bugs.launchpad.net/juju/+bug/1738209

### CODE REVIEW

File: state/txns.go [hunk 1]
```diff
@@ -53,10 +55,32 @@ type multiModelRunner struct {
...
 }
 
+func shortStack() string {
+	var rv []string
+	for _, line := range strings.Split(string(debug.Stack()), "\n") {
+		if len(line) > 0 && line[0] == '\t' {
+			rv = append(rv, strings.Split(line[1:], " ")[0])
+		}
+	}
+	// We definitely have at least 3 objects in rv - debug.Stack(), this function and the one that called it.
```
Comment (jameinel) [inline state/txns.go]: // Trim off debug.Stack() and shortStack() and the caller of shortStack from the stack trace.
I wonder if we actually want to trim the caller. RunTransaction() seems relevant to someone debugging it.

**Comment by jameinel** [comment]
This is the line I had been talking about:
https://github.com/juju/txn/blob/master/txn.go#L185

We should already not be trying to actually mutate the database if len(ops) == 0.
We could add logging there as well if we really want to push this all the way up into callers that have no operations to produce.

**Comment by jameinel** [comment]
(callers with len(ops) == 0 should be returning ErrNoOperations from their method)

**Comment by wupeka** [comment]
!!test!!

**Comment by jameinel** [comment]
$$merge$$

**Pull request merged**